### PR TITLE
Add plug collection

### PR DESCRIPTION
**Pull request by gabriel-samfira**
This PR adds the ability for providers to consume storage that requires the target machine agent to initialize the storage before it becomes available locally.

One such example is ```iSCSI```, which at the moment is the only type of attachable storage in the Oracle OCI cloud (former Bare Metal Cloud - bmcs).

The OCI cloud allows creating volumes, and attaching them to instances, but there is still a need to log into the iscsi target using the connection info and credentials returned as part of the GetAttachment() API call. There is no agent inside the instances that does this automatically, so it is left up to the juju machine agent to run those commands.

This change adds a new collection in the database called ```volumeattachmentplug``` which is watched by machine agents, and which prompts the machine agent, to fetch the volume plug info set by the provider during volume creation/attachment, and do any initialization needed.

After the machine agent finishes initializing the volume attachment connection, it will send back the block device information relevant to that attachment (device name, WWN, serial number, etc). This means that the cloud provider code does not need to set the device name. It is automatically discovered by the storage plug.

This change should *not* require an update in any of the other providers. The lack of "plug info" should just be a noop for the other providers. Also, it should not require any upgrade steps for existing environments. The lack of a "storage plug" will simply mean that the old code path is used.

This change means that any provider can use external storage sources like iSCSI/Ceph that require some extra steps to be run on the machine itself before being able to access the block device.

**NOTE**: depends on: https://github.com/juju/description/pull/41

QA steps:

On OCI:

- Boostrap new controller: SUCCESS
- deploy workload with storage attachment: SUCCESS
- detach storage from unit: SUCCESS
- re-attach storage to unit: SUCCESS

On AWS:
- Boostrap new controller: SUCCESS
- deploy workload with storage attachment: SUCCESS
- detach storage from unit: SUCCESS
- re-attach storage to unit: SUCCESS

### CODE REVIEW

**Comment by jameinel** [comment]
[Volume Plugs](https://trello.com/c/FX913MDz/88-volume-plugs)

**Comment by howbazaar** [comment]
When adding collections, we need to make sure that we consider upgrades for when docs are assumed to be there, and also model migration.

**Comment by howbazaar** [comment]
@gabriel-samfira are the storage plug documents persistent or transitory?

If they are just transitory, then we could make model migration pre-checks ensure that there are none pending. However if they are persistent, we'll need to add the ability to record them in the description package and handle them in the state/migration import and export files.

**Comment by gabriel-samfira** [comment]
@howbazaar As it is currently written, the code shortcuts to the next step if storage plugs are not there, assuming we upgraded from a previous version that did not have (or need) them. Up until this point, attaching disks to instances was transparent for the instaces, on all providers, so there was no need to notify the machine agents that they needed to do any extra steps.

The information in the plugs document persists only as long as the attachment exists. That info gets deleted when the disk gets detached.

**Comment by jameinel** [comment]
This is a reasonably interesting PR adding new fields to description:
https://github.com/juju/description/pull/32/files

File: storage/plugs/iscsi/iscsi.go [hunk 1]
```diff
@@ -0,0 +1,369 @@
+package iscsi
```
Comment (manadart) [inline storage/plugs/iscsi/iscsi.go]: All these new files need copyright notices.

File: storage/plugs/iscsi/iscsi.go [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+		return storage.BlockDevice{}, errors.Trace(err)
+	}
+	return getHardwareInfo(devName)
+}
+
+func (i *iscsiConnectionInfo) unplug() error {
+	if err := i.logout(); err != nil {
+		return errors.Trace(err)
+	}
+	return i.delete()
```
Comment (manadart) [inline storage/plugs/iscsi/iscsi.go]: Probably trace this too.

File: storage/volume.go [hunk 3]
```diff
@@ -5,6 +5,13 @@ package storage
 
 import "gopkg.in/juju/names.v2"
 
+type DeviceType string
+
+var (
+	DiskTypeLocal DeviceType = "local"
```
Comment (jameinel) [inline storage/volume.go]: This is mixing the overall "type DeviceType" with all prefixes that are "DiskType*" we probably want to be consistent here.

File: worker/storageprovisioner/volume_events.go [hunk 4]
```diff
@@ -54,6 +55,122 @@ func volumesChanged(ctx *context, changes []string) error {
...
+		}
+	}
+
+	results, err := ctx.config.Volumes.SetVolumeAttachmentPlugBlockInfo(volumeAttachmentPlugs)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	for _, result := range results {
+		if result.Error != nil {
+			return errors.Errorf("failed to publish bloci info to state: %s", result.Error)
```
Comment (manadart) [inline worker/storageprovisioner/volume_events.go]: Typo.

File: worker/provisioner/provisioner_task.go [hunk 5]
```diff
@@ -1263,11 +1263,17 @@ func volumesToAPIserver(volumes []storage.Volume) []params.Volume {
 func volumeAttachmentsToAPIserver(attachments []storage.VolumeAttachment) map[string]params.VolumeAttachmentInfo {
 	result := make(map[string]params.VolumeAttachmentInfo)
 	for _, a := range attachments {
+		var plugInfo *params.VolumeAttachmentPlanInfo
```
Comment (manadart) [inline worker/provisioner/provisioner_task.go]: This looks like the only remaining "plug" reference in code.

File: apiserver/facades/agent/storageprovisioner/storageprovisioner.go [hunk 6]
```diff
@@ -1045,6 +1173,30 @@ func (s *StorageProvisionerAPIv3) oneVolumeBlockDevice(
...
+	// TODO(gsamfira): revisit this
+	planCanAccess, err := s.getMachineAuthFunc()
+	if err != nil && !errors.IsNotFound(err) {
+		return state.BlockDeviceInfo{}, err
+	}
+	var blockDeviceInfo state.BlockDeviceInfo
+
+	volumeAttachmentPlan, err := s.oneVolumeAttachmentPlan(id, planCanAccess)
+	if err != nil {
+		// This is optional. It should not err out
```
Comment (jameinel) [inline apiserver/facades/agent/storageprovisioner/storageprovisioner.go]: Being clearer about what "this" is is probably useful.
Is it the BlockDeviceInfo that is optional?

File: apiserver/facades/agent/uniter/mock_test.go [hunk 7]
```diff
@@ -19,6 +19,7 @@ type fakeStorage struct {
 	storageInstance        func(names.StorageTag) (state.StorageInstance, error)
 	storageInstanceVolume  func(names.StorageTag) (state.Volume, error)
 	volumeAttachment       func(names.Tag, names.VolumeTag) (state.VolumeAttachment, error)
+	volumeAttachmentPlan   func(names.Tag, names.VolumeTag) (state.VolumeAttachmentPlan, error)
```
Comment (jameinel) [inline apiserver/facades/agent/uniter/mock_test.go]: I'm guessing that at this point there is too much history and other code involved to do this as a mockgen based mock. Is that your feeling?

File: apiserver/common/storagecommon/mock_test.go [hunk 8]
```diff
@@ -41,6 +42,11 @@ func (s *fakeStorage) VolumeAttachment(m names.Tag, v names.VolumeTag) (state.Vo
 	return s.volumeAttachment(m, v)
 }
 
+func (s *fakeStorage) VolumeAttachmentPlan(host names.Tag, volume names.VolumeTag) (state.VolumeAttachmentPlan, error) {
```
Comment (jameinel) [inline apiserver/common/storagecommon/mock_test.go]: Is this method still being used? When I tried to look around I thought it was mostly changed to VolumeAttachmentPlans. Is there a reason we need both? I guess I'm a bit confused why one takes a host and a volume, and the other takes only the volume.

File: storage/plans/iscsi/iscsi.go [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+	cmd := strings.Join(params, " ")
+	execParams := exec.RunParams{
+		Commands: cmd,
+	}
+	resp, err := exec.RunCommands(execParams)
+
+	return resp, err
+}
+
+func getHardwareInfo(name string) (storage.BlockDevice, error) {
```
Comment (jameinel) [inline storage/plans/iscsi/iscsi.go]: I think it would make sense to split this into 2 separate functions for testing purposes. One would actually call an external command, and then just return an io.Reader that you could use.
Then you can use a bytes.Buffer to test the function that parses the result, and easily feed that function test input so we can make sure it copes with any edge cases that we find in the wild.

File: storage/plans/iscsi/iscsi.go [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+}
+
+// addTarget adds the iscsi target config
+func (i *iscsiConnectionInfo) addTarget() error {
+	newNodeParams := []string{
+		"iscsiadm", "-m", "node",
+		"-o", "new",
+		"-T", i.iqn,
+		"-p", i.portal()}
+	result, err := runCommand(newNodeParams)
```
Comment (jameinel) [inline storage/plans/iscsi/iscsi.go]: Given that these take mostly the same parameters, I wonder if it is worth having a helper function.
That said, it can also be nice to see the whole command that is being executed in one spot, rather than having to look around for which parameters are/aren't being passed.
Probably the clarity of what full command is being run wins out in the end.
They seem to all take -m, -o, -T, but not all take -p/-n/-v. Would it be clearer as a parameter struct?

runISCSICommand(iscsiParameter{
  Mode: "node",
  Operation: "update",
  T?: i.iqn,
  N?: "node.startup",
  V?: "automatic",
  Portal: i.portal(),
etc.

Think about it.

File: apiserver/common/storagecommon/mock_test.go (hunk 8 above)
Comment (gabriel-samfira) [reply apiserver/common/storagecommon/mock_test.go]: Yes. See:

https://github.com/gabriel-samfira/juju/blob/add-storage-plugs/apiserver/facades/agent/storageprovisioner/storageprovisioner.go#L1127

One gets all attachments for a host. The other gets a particular attachment for one volume.

File: apiserver/facades/agent/storageprovisioner/storageprovisioner.go (hunk 6 above)
Comment (gabriel-samfira) [reply apiserver/facades/agent/storageprovisioner/storageprovisioner.go]: Will update to be more clear. It refers to the volume attachment plans. They are not mandatory. If missing, defaults to old code path.

File: apiserver/facades/agent/uniter/mock_test.go (hunk 7 above)
Comment (gabriel-samfira) [reply apiserver/facades/agent/uniter/mock_test.go]: If I do the mockgen thing for this, I'd have to refactor for every other bit of test code here, which is something I do not want to do as part of this PR.

File: storage/plans/iscsi/iscsi.go (hunk 10 above)
Comment (gabriel-samfira) [reply storage/plans/iscsi/iscsi.go]: So, the whole module is around 400 lines of code, which essentially does a simple thing like running a bunch of exec commands. Module is pretty self contained and simple. I would like to avoid over engineering it. I avoided creating a struct, because the task seemed simple enough and clear enough to get away with a simple ```[]string{}```. Anything more complicated than that seemed overkill for the current purpose.

File: storage/plans/iscsi/iscsi.go (hunk 9 above)
Comment (gabriel-samfira) [reply storage/plans/iscsi/iscsi.go]: I think we can just get away with mocking out ```runCommand``` and returning whatever output we want?

File: apiserver/facades/agent/storageprovisioner/storageprovisioner.go [hunk 11]
```diff
@@ -1098,6 +1226,32 @@ func (s *StorageProvisionerAPIv3) oneVolumeBlockDevice(
 	if err != nil {
 		return state.BlockDeviceInfo{}, err
 	}
+	// TODO(gsamfira): revisit this
```
Comment (manadart) [inline apiserver/facades/agent/storageprovisioner/storageprovisioner.go]: Does this still need to be revisited?

File: apiserver/facades/client/storage/mock_test.go [hunk 12]
```diff
@@ -99,6 +101,16 @@ func (st *mockStorageAccessor) VolumeAttachment(m names.Tag, v names.VolumeTag)
 	return st.volumeAttachment(m, v)
 }
 
+func (st *mockStorageAccessor) VolumeAttachmentPlan(host names.Tag, volume names.VolumeTag) (state.VolumeAttachmentPlan, error) {
+	// st.MethodCall(st, "VolumeAttachmentPlan", host, volume)
```
Comment (manadart) [inline apiserver/facades/client/storage/mock_test.go]: Might as well remove these.

File: state/migration_export.go [hunk 13]
```diff
@@ -1846,6 +1889,26 @@ func (e *exporter) readVolumeAttachments() (map[string][]volumeAttachmentDoc, er
...
+	var doc volumeAttachmentPlanDoc
+	var count int
+	iter := coll.Find(nil).Iter()
+	defer iter.Close()
+	for iter.Next(&doc) {
+		result[doc.Volume] = append(result[doc.Volume], doc)
+		count++
+	}
+	if err := iter.Close(); err != nil {
+		return nil, errors.Annotate(err, "failed to read volumes attachment plans")
```
Comment (manadart) [inline state/migration_export.go]: "volume attachment plans"

File: apiserver/common/storagecommon/blockdevices.go [hunk 14]
```diff
@@ -32,6 +32,7 @@ func MatchingBlockDevice(
 	blockDevices []state.BlockDeviceInfo,
 	volumeInfo state.VolumeInfo,
 	attachmentInfo state.VolumeAttachmentInfo,
+	planBlockInfo state.BlockDeviceInfo,
```
Comment (manadart) [inline apiserver/common/storagecommon/blockdevices.go]: This method has no pre-existing tests, but can we get the ball rolling with one that tests the added path?

File: api/storageprovisioner/provisioner.go [hunk 15]
```diff
@@ -120,6 +120,12 @@ func (st *State) WatchVolumeAttachments(scope names.Tag) (watcher.MachineStorage
 	return st.watchAttachments("WatchVolumeAttachments", scope, apiwatcher.NewVolumeAttachmentsWatcher)
 }
 
+// WatchVolumeAttachmentPlans watches for changes to volume attachments
+// scoped to the entity with the tag passed to NewState.
+func (st *State) WatchVolumeAttachmentPlans(scope names.Tag) (watcher.MachineStorageIdsWatcher, error) {
```
Comment (manadart) [inline api/storageprovisioner/provisioner.go]: This and the other additions to the package could use tests. There are precedents in provisioner_test.go.

**Comment by jameinel** [comment]
!!build!!

**Comment by jameinel** [comment]
Joe and I discussed this, and we think it is certainly past the bar for what we need from code coverage. We'll certainly want to keep an eye on this, and when we come to issues, we should feel like refactoring and increasing coverage is always a good thing.

Thanks for all your work on this. When I just checked, it seemed the Windows tests were failing, but I couldn't get the link to the actual failure. So for tonight I'll submit it, but if the windows tests fail, we'll need them fixed before we can land it. (As of last week we moved windows to be part of the pre-commit tests).

$$merge$$

**Comment by gabriel-samfira** [comment]
!!build!!

**Comment by gabriel-samfira** [comment]
!!build!!

**Comment by jameinel** [comment]
!!build!!

**Comment by jameinel** [comment]
The last failure was the script failing to download go2xunit, which is clearly not the fault of the patch.

**Comment by jameinel** [comment]
I can't actually find the failure in: http://ci.jujucharms.com/job/make-check-juju/268/consoleFull other than it finishing with:
make: *** [check] Error 2
Makefile:70: recipe for target 'check' failed
2018-09-25 04:20:32+00:00 ran make check
+ cleanup
+ exit_code=2

I do see several complaints from debconf:
debconf: unable to initialize frontend: Dialog
debconf: (Dialog frontend will not work on a dumb terminal, an emacs shell buffer, or without a controlling terminal.)
debconf: falling back to frontend: Readline
debconf: unable to initialize frontend: Readline
debconf: (This frontend requires a controlling tty.)
debconf: falling back to frontend: Teletype
dpkg-preconfigure: unable to re-open stdin: 

and some:
warning: ignoring symlink /home/ubuntu/go/src/github.com/juju/juju/vendor/github.com/prometheus/procfs/fixtures/self

but all the go fmt, go vet, etc all seem happy.

I see some warnings, some test stipple (TLS Handshake error is rather common)
The only one that maybe is something is:
testing: warning: no tests to run

But that looks more like a warning than a failure.
--- PASS: TestPackage (0.01s)
PASS
ok  	github.com/juju/juju/service/common	0.021s
testing: warning: no tests to run
PASS
ok  	github.com/juju/juju/service/common/testing	0.011s [no tests to run]
=== RUN   Test


Anyone else see something I missed?

**Comment by manadart** [comment]
@jameinel @gabriel-samfira In Brussels last week, the CI linting and sundries were updated for Go 1.11.

I'd hazard a guess that 1.10 gfmt'd code now makes it unhappy. I know Gabriel downgraded to 1.10 to accommodate it previously.

**Comment by gabriel-samfira** [comment]
upgraded the golang snap back to 1.11. Formatted and pushed. Let's see what happens.

**Comment by gabriel-samfira** [comment]
!!build!!

**Comment by manadart** [comment]
!!build!!

**Comment by manadart** [comment]
```
storage/plans/iscsi/iscsi.go:318: Annotatef format %s has arg result of wrong type *github.com/juju/juju/vendor/github.com/juju/utils/exec.ExecResponse
storage/plans/iscsi/iscsi.go:334: Annotatef format %s has arg result of wrong type *github.com/juju/juju/vendor/github.com/juju/utils/exec.ExecResponse
storage/plans/iscsi/iscsi.go:350: Annotatef format %s has arg result of wrong type *github.com/juju/juju/vendor/github.com/juju/utils/exec.ExecResponse
```

**Comment by manadart** [comment]
$$merge$$
